pkg/block/metadata/hash.go

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package metadata
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// HashKind indicates what type of hash it is.
+type HashKind string
+
+const (
+	// SHA256Kind shows that SHA256 has been used to generate the hash.
+	SHA256Kind HashKind = "SHA256"
+)
+
+// ObjectHash stores the hash of an object in the object storage.
+type ObjectHash struct {
+	Kind  HashKind
+	Value string
+}
+
+// Equal returns true if two hashes are equal.
+func (oh *ObjectHash) Equal(other *ObjectHash) bool {
+	return oh.Value == other.Value
+}
+
+// CalculateHash calculates the hash of the given type.
+func CalculateHash(p string, k HashKind) (ObjectHash, error) {
+	switch k {
+	case SHA256Kind:
+		f, err := os.Open(p)
+		if err != nil {
+			return ObjectHash{}, errors.Wrap(err, "opening file")
+		}
+		defer f.Close()
+
+		h := sha256.New()
+
+		if _, err := io.Copy(h, f); err != nil {
+			return ObjectHash{}, errors.Wrap(err, "copying")
+		}
+
+		return ObjectHash{
+			Kind:  SHA256Kind,
+			Value: hex.EncodeToString(h.Sum(nil)),
+		}, nil
+	default:
+		panic(fmt.Sprintf("hash kind %v is not supported; fill a bug, please", k))
+	}
+}
+
+// GetHashDir calculates the hashes of the given kind of all files inside of dir.
+// Ignores the given files.
+func GetHashDir(dir string, k HashKind, ignore ...string) (map[string]ObjectHash, error) {
+	ret := map[string]ObjectHash{}
+	if err := filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		cleanPath := filepath.Join(strings.TrimPrefix(filepath.Dir(filepath.Clean(path)), dir), info.Name())
+
+		for _, v := range ignore {
+			if cleanPath == v || (!strings.Contains(v, "/") && filepath.Base(v) == info.Name()) {
+				if info.IsDir() {
+					return filepath.SkipDir
+				}
+				return nil
+			}
+		}
+
+		if info.IsDir() {
+			return nil
+		}
+
+		h, err := CalculateHash(path, k)
+		if err != nil {
+			return errors.Wrapf(err, "calculating hash of %s", dir)
+		}
+		ret[strings.TrimPrefix(cleanPath, "/")] = h
+		return nil
+	}); err != nil {
+		return ret, errors.Wrap(err, "walk")
+	}
+
+	return ret, nil
+}
+
+// NominalIgnoredFiles is a list of files that you want to typically ignore
+// when calculating the hashes of a TSDB block.
+var NominalIgnoredFiles = []string{DeletionMarkFilename, MetaFilename}
```

Comment:
Why?

Reply:
Well, the deletion mark could actually just be removed from here since it will never appear while a block is being produced. However, I think the `meta.json` file should be definitely excluded because there will be confusion because the hash will most likely change *after* adding it to `m.Thanos.Hashes`.

Reply:
I mean, I would just rename it tbh or just use `meta.json` explicitly somewhere. `Nominal` does not mean anything to me :hugs: 

But overall it makes sense to skip hashing meta.json, it would hard... to achieve (: